features/netsocket/SocketStats.h

```diff
@@ -0,0 +1,115 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SOCKET_STATS_H
+#define SOCKET_STATS_H
+
+#include "platform/mbed_stats.h"
+#include "platform/SingletonPtr.h"
+#include "platform/PlatformMutex.h"
+#include "netsocket/Socket.h"
+#include "SocketAddress.h"
+#include "hal/ticker_api.h"
+
+#ifndef MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT
+#define MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT      10
+#endif
+
+typedef enum {
+    SOCK_CLOSED,                    /**< Socket is closed and does not exist anymore in the system */
+    SOCK_OPEN,                      /**< Socket is open, but not associated to any peer address */
+    SOCK_CONNECTED,                 /**< Socket is associated to peer address, either by connect() or sendto()/recvfrom() calls */
+    SOCK_LISTEN,                    /**< Socket is listening for incoming connections */
+} socket_state;
+
+typedef struct {
+    Socket *reference_id;           /**< Used for identifying socket */
+    SocketAddress peer;             /**< Last associated peername of this socket (Destination address) */
+    socket_state state;             /**< State of this socket */
+    nsapi_protocol_t proto;         /**< Specifies a protocol used with socket */
+    size_t sent_bytes;              /**< Data sent through this socket */
+    size_t recv_bytes;              /**< Data received through this socket */
+    us_timestamp_t last_change_tick;/**< osKernelGetTick() when state last changed */
+} mbed_stats_socket_t;
+
+/**  SocketStats class
+ *
+ *   Class to get the network socket statistics
+ */
+class SocketStats {
+public:
+
+    /** Create an socket statictics object
+     *
+     *  Applicatiopn users must not create class object.
+     *  Class object will be created by entities reporting network statistics.
+     *  Application can fetch network statistics using static `mbed_stats_socket_get_each` API
+     *  without creatin an object.
```

Comment:
```suggestion
     *  without creating an object.
```

Reply:
Sorry for the large blob of comments...Github doens't support multi-line edits yet, all very minor nits

Reply:
@kegilbert - Github allows you to merge also one by one :-) Will squash them all